x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportPutTrainedModelAction.java

```diff
@@ -0,0 +1,190 @@
+import org.elasticsearch.xpack.core.ml.utils.ExceptionsHelper;
+import org.elasticsearch.xpack.ml.inference.persistence.TrainedModelProvider;
+
+import java.io.IOException;
+import java.time.Instant;
+import java.util.List;
+
+import static org.elasticsearch.xpack.core.ClientHelper.ML_ORIGIN;
+import static org.elasticsearch.xpack.core.ClientHelper.executeAsyncWithOrigin;
+
+public class TransportPutTrainedModelAction extends TransportMasterNodeAction<Request, Response> {
+
+    private final TrainedModelProvider trainedModelProvider;
+    private final XPackLicenseState licenseState;
+    private final NamedXContentRegistry xContentRegistry;
+    private final Client client;
+
+    @Inject
+    public TransportPutTrainedModelAction(TransportService transportService, ClusterService clusterService,
+                                          ThreadPool threadPool, XPackLicenseState licenseState, ActionFilters actionFilters,
+                                          IndexNameExpressionResolver indexNameExpressionResolver, Client client,
+                                          TrainedModelProvider trainedModelProvider, NamedXContentRegistry xContentRegistry) {
+        super(PutTrainedModelAction.NAME, transportService, clusterService, threadPool, actionFilters, Request::new,
+            indexNameExpressionResolver);
+        this.licenseState = licenseState;
+        this.trainedModelProvider = trainedModelProvider;
+        this.xContentRegistry = xContentRegistry;
+        this.client = client;
+    }
+
+    @Override
+    protected String executor() {
+        return ThreadPool.Names.SAME;
+    }
+
+    @Override
+    protected Response read(StreamInput in) throws IOException {
+        return new Response(in);
+    }
+
+    @Override
+    protected void masterOperation(Task task,
+                                   PutTrainedModelAction.Request request,
+                                   ClusterState state,
+                                   ActionListener<Response> listener) {
+        try {
+            request.getTrainedModelConfig().ensureParsedDefinition(xContentRegistry);
+            request.getTrainedModelConfig().getModelDefinition().getTrainedModel().validate();
+        } catch (IOException ex) {
+            listener.onFailure(ExceptionsHelper.badRequestException("Failed to parse definition for [{}]",
+                ex,
+                request.getTrainedModelConfig().getModelId()));
+            return;
+        } catch (ElasticsearchException ex) {
+            listener.onFailure(ExceptionsHelper.badRequestException("Definition for [{}] has validation failures.",
+                ex,
+                request.getTrainedModelConfig().getModelId()));
+            return;
+        }
+
+        TrainedModelConfig trainedModelConfig = new TrainedModelConfig.Builder(request.getTrainedModelConfig())
+            .setVersion(Version.CURRENT)
+            .setCreateTime(Instant.now())
+            .setCreatedBy("user")
```

Comment:
I thought created by would be the users name (if security is enabled) but now I realise not. Maybe `api` rather than `user`. Or `api_user`. Our included models are `createdBy: ml_admin`

Reply:
`api_user` seems good to me. 

Yeah, I don't think we should include the actual user name for security reasons. It seems like a potential to leak info. Somebody could have permissions to read trained models, but not permission to see all configured users.

Reply:
Maybe it would be best to match what we did for system created annotations. Otherwise we’re just creating a plethora of imaginary users and eventually someone will ask why they’re all different.

Reply:
@droberts195 for our internal created annotations, we use `XPackUser.NAME` which is `_xpack`. I am fine switching it to that in a subsequent PR.